minor: Split BracesFixer

julienfalque opened the pull request:
Fixes #823, #776, #4950, #5115 and #6519; replaces #4721.

Extracted:
- [x] #5751
- [x] #5932
- [x] #5960
- [x] #6399
- [x] #6452
- [x] #6458

keradus commented on src/Fixer/Basic/BracesFixer.php:
```diff
@@ -124,33 +136,53 @@ public function bar($baz)
-        $this->fixControlContinuationBraces($tokens);
-        $this->fixSpaceAroundToken($tokens);
-        $this->fixDoWhile($tokens);
+        for ($index = \count($tokens) - 1; $index > 0; --$index) {
```
questions:
- why adding those spaces before running the proxy fixers?
- if someone would be using the proxied fixers directly, without BracesFixer, who would add those spaces?

julienfalque replied:
This is something I added as a TODO task in #4885. There is a PR (#4435) that could be a candidate replacement so I didn't implement one myself for now.

I moved the task here.

keradus replied:
nice!

julienfalque replied:
Just noticed I didn't actually add a TODO task to #4885 so I did it :smile:

keradus commented on tests/Fixer/PhpUnit/PhpUnitExpectationFixerTest.php:
```diff
@@ -305,7 +305,7 @@ function testFnc%d()
         }
 '
-;
+        ;
```
why this change? is it expected?

julienfalque replied:
Yep, the new `StatementIndentationFixer` reindents all lines unless they are part of a multiline statement. It considers `;` to *not* be part of the expression, e.g.:

```diff
-  $foo =
-            true
-             || false
-            ;
+$foo =
+          true
+           || false
+;
```
Reindenting the lines that are part of a statement should be another rule IMO, out of the scope of this PR.

keradus replied:
not sure if it's out of scope - like one would say it's a change of behaviour of existing fixer.

is this new behaviour (the extra spaces) are improving the rule or not?

keradus replied:
looking good

keradus commented on src/Fixer/Basic/BracesFixer.php:
```diff
@@ -178,892 +210,45 @@ protected function createConfigurationDefinition()
-            T_ELSE,
-            T_ELSEIF,
-            T_FINALLY,
+        return [
```
I wonder if we should add some integration test for relations between those newly created fixers...

julienfalque replied:
Maybe some priority tests indeed, otherwise I think tests in `BracesFixerTest` should be fine

keradus replied:
but on line v3, we gonna drop whole `BracesFixerTest` file, right?
so, we should have the replacement for it on v2.x as well. good we did not drop the cases in `BracesFixerTest` yet, but the new ones should be added as well

julienfalque replied:
I added some integration tests.

keradus commented on src/Fixer/Whitespace/StatementIndentationFixer.php:
```diff
@@ -0,0 +1,600 @@
+    /**
+     * @param bool $bracesFixerCompatibility
+     */
+    public function __construct($bracesFixerCompatibility = false)
```
Please remove constructor parameter from this and other fixers.
Fixers should become services and receive only dependencies via constructor.

This look like configuration (the option should state what the trigger changes in behaviour of fixer, not if it's compat mode with old fixer or not)

julienfalque replied:
I used these options to drop behaviors from `braces` that were too specific and opinionated IMO and that I didn't want to keep. The new fixers don't have those behaviors at all, unless executed via `braces` for backward compatibility.

Do you want to keep the `braces` specific behavior with the new fixers?

keradus replied:
what are the `specific behaviors` of `braces`?
can't understand it by looking at `bracesFixerCompatibility` name :(
I think we should decide one by one.

Same time, fixer is public, please no constructor parameter like this.
If we would decide that sth is too weird and we keep it as compatibility only and not as a proper configuration option, then create a method like `/** @internal */ enableBracesCompatibilityMode(): void {...}`

julienfalque replied:
Here is the list of `braces` specific behaviors, with test failures from `BracesFixerTest` that cover them.

<details><summary>1. In <code>control_structure_braces</code>: when a control structure has no braces and contains an alternative syntax control structure, do not add braces on the outer control structure</summary>

```diff
-<?php if ($a) foreach ($b as $c): ?> X <?php endforeach; ?>
+<?php if ($a) {
+    foreach ($b as $c): ?> X <?php endforeach;
+} ?>
```

Maybe we could improve this by using the alternative syntax on the outer control structure (but out of scope here I'd say):

```diff
-<?php if ($a) foreach ($b as $c): ?> X <?php endforeach; ?>
+<?php if ($a): foreach ($b as $c): ?> X <?php endforeach; endif; ?>
```

</details>

<details><summary>2. In <code>blank_lines_inside_block</code>: when a curly braces block starts with blank lines followed by a comment, don't remove the blank lines</summary>

```diff
 <?php
 if (true) {
-
     //  The blank line helps with legibility in nested control structures
     if (true) {
         // if body
     }
 
     // if body
 }
```

</details>

<details><summary>3. In <code>statement_indentation</code>: when a comment looks like commented-out code, don't indent it</summary>

```diff
 <?php
 function foo()
 {
     $a = 1;
-//    if ($a === 'bar') {
-//        return [];
-//    }
+    //    if ($a === 'bar') {
+    //        return [];
+    //    }
     // we will return sth
     return $a;
 }
```

</details>

<details><summary>4. In <code>statement_indentation</code>: when a single line comment is part of a multiline message, don't indent it</summary>

```diff
 <?php
 function foo()
 {
     $bar = 1;                   // multiline ...
-                                // ... comment
+    // ... comment
     $baz  = 2;                  // next comment
 }
```

</details>

<details><summary>5. In <code>statement_indentation</code>: assume code that is not inside any curly braces block is correctly indented</summary>

```diff
 <?php
-    if ($test) { //foo
-        echo 1;
-    }
+if ($test) { //foo
+    echo 1;
+}
```

</details>

keradus replied:
ref https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/4884#issuecomment-1122604242

I believe we should not change code to START using an alternative syntax
about 1&2 @julienfalque , I'm OK to keep it nice way in new fixers, as we merged them, and actually _any_ way in braces. let's get this deprecation merged asap, even if maybe code is not perfect - I will name it my mistake to aim to have it nice here (eg no config passed to constructor). We should aim to limit the impact of code-smell to other fixers even if it partially affect braces fixer in way we didn't wanted initially, and then we should bump major version to get rid of braces rule.

julienfalque replied:
> let's figure out how to minify remaining work to have this PR merged

I'm preparing a new PR to extract parts from this one.

> then let's open PR for v4 cleanup of legacy `Braces` fixer and it's spill-over (`$bracesFixerCompatibility`) to fresh rules

This work already exists but will need an epic rebase at some point: #4885

julienfalque replied:
`$bracesFixerCompatibility` dropped for `control_structure_braces` and `blank_lines_inside_block`.

wshaver commented:
Hoping for progress on this for comments and php open tag issues. Details here:
https://github.com/FriendsOfPHP/PHP-CS-Fixer/issues/5019#issuecomment-707372126

shakaran commented:
@julienfalque could you update the status of this draft? How we can help you to speed up?

julienfalque commented:
@shakaran We need to make a decision regarding https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/4884#discussion_r397103973. Also, I don't have much time to work on this currently so any contribution is welcome :)

julienfalque commented:
Rebased and added test for `AlternativeSyntaxAnalyzer`.

Now everything in `BracesFixer` is delegated to other fixers except one thing: ensuring there is a single space before the `use` keyword in anonymous functions. For that I propose to replace `SingleSpaceAfterConstructFixer` with a new `SpacingAroundConstructFixer` that would take care of it. What do you think @keradus @kubawerlos @SpacePossum @localheinz?

julienfalque commented:
I started splitting this PR into multiple smaller ones. First one is #5751.

keradus commented:
#5751 merged . I believe we can extract next element ;)

julienfalque commented:
Here it is: #5932 😄

keradus commented:
@julienfalque , https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/5960 got merged. Can you update the PR? is there any open topic here, that is left? or maybe a decision that is awaiting to be made?

julienfalque commented:
Next step: #6452.

julienfalque commented:
Next step: #6508 😄

julienfalque commented:
This PR is now small enought to be reviewed directly IMO.

SpacePossum commented on src/Fixer/Whitespace/NoExtraBlankLinesFixer.php:
```diff
@@ -441,13 +441,7 @@ private function removeEmptyLinesAfterLineWithTokenAt(int $index): void
                 continue;
             }
 
-            $pos = strrpos($content, "\n");
```
is this a clean up or can we maybe add a utest for this change to NoExtraBlankLinesFixerTest as well?

julienfalque replied:
This is only a refactoring that shouldn't change behavior.

keradus reviewed (approved):
So big effort, in this and extracted PRs!
big, big kudos, @julienfalque !

TomasVotruba commented:
Thanks for the split :+1: It will be helfpul in narrowing down changes and bugs!

It seems this split broke indent of catch/switch. How should I setup the rules to keep the original behavior?

```diff
-switch ($simple) {\n
+switch($simple) {\n
```

```diff
-} catch (FirstThrowableType $e) {\n
+} catch(FirstThrowableType $e) {\n
```

```diff
 new class($a, $b, $c) extends SomeExtendedClass implements\n
     \ArrayAccess,\n
     \Countable {\n
-        // Body\n
-    };\n
+    // Body\n
+};\n
```

TomasVotruba commented:
Allright, add `SingleSpaceAfterConstructFixer` solves first 2 cases :+1: :slightly_smiling_face:

julienfalque commented:
@TomasVotruba Third case looks expected though.

TomasVotruba commented:
@julienfalque I see people complaining PSR12 set is now broken :) How can I make it 3.9 compatible?

julienfalque commented:
@TomasVotruba Please submit an issue with details regarding what is broken.

kayw-geek commented:
@julienfalque 
I found a bug in the following:

```PHP
<?php
    class Foo
    {
    public function test()
    { 
    }
    }
```
it will be fixed before v3.10.0, but it doesn't fix now.

This PR is big change, so I not found where line has a problem

kayw-geek commented:
> @julienfalque I found a bug in the following:
> 
> ```
> <?php
>     class Foo
>     {
>     public function test()
>     { 
>     }
>     }
> ```
> 
> it will be fixed before v3.10.0, but it doesn't fix now.
> 
> This PR is big change, so I not found where line has a problem

`src/Fixer/Basic/CurlyBracesPositionFixer.php`

keradus commented:
@kayw-geek , would you mind to open new issue and link this PR? not so many ppl are watching the closed issues/PRs

XBrayanX commented:
First of all, thank you very much for your great work. 
My Question is, with the new changes, is there a rule to keep empty spaces inside control structures? It is mentioned in #4487

julienfalque commented:
No, the rules created during this refactoring only cover existing behaviour of `braces`.

XBrayanX commented:
But if you will pay attention to comment #4487? This comment is already 3 years old, please, this functionality is necessary to improve the code's readability.

keradus commented:
I don't believe we ever provided this feature, and it's not related to split of Braces fixer. if you would like to provide such feature, please show it's needed by given community [eg some big open source project/framework] and propose a PR for it, thanks !
